daemon/emer-daemon.c

```diff
@@ -1123,6 +1133,178 @@ set_max_bytes_buffered (EmerDaemon *self,
+             date);
+
+  save_aggregate_timers_to_tally (self, date, now_monotonic_us);
+  emer_aggregate_tally_iter (priv->aggregate_tally,
+                             date,
+                             EMER_TALLY_ITER_FLAG_DELETE,
+                             buffer_aggregate_event_to_queue,
+                             self);
+
+  /* Monthly events */
+  now = g_date_time_new_now_local ();
+  if (month_changed (now, priv->current_aggregate_tally_date))
+    {
+      g_autofree char *month_date = NULL;
+
+      month_date = g_date_time_format (priv->current_aggregate_tally_date,
+                                       "%Y-%m");
+
+      g_message ("Buffering monthly aggregate events from %s to submission queue",
+                 month_date);
+
+      save_aggregate_timers_to_tally (self, month_date, now_monotonic_us);
+      emer_aggregate_tally_iter (priv->aggregate_tally,
+                                 month_date,
+                                 EMER_TALLY_ITER_FLAG_DELETE,
+                                 buffer_aggregate_event_to_queue,
+                                 self);
+    }
+
+  split_aggregate_timers (self, now_monotonic_us);
+
+  schedule_next_midnight_tick (self);
+
+  return G_SOURCE_REMOVE;
+}
+
+static void
+schedule_next_midnight_tick (EmerDaemon *self)
+{
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+  g_autoptr (GDateTime) next_midnight = NULL;
+  g_autoptr (GDateTime) next_day = NULL;
+  g_autoptr (GDateTime) now = NULL;
+  gint64 time_to_next_midnight_us;
+
+  now = g_date_time_new_now_local ();
+  next_day = g_date_time_add_days (now, 1);
+  next_midnight = g_date_time_new_local (g_date_time_get_year (next_day),
+                                         g_date_time_get_month (next_day),
+                                         g_date_time_get_day_of_month (next_day),
+                                         0, 0, 0.0);
+
+  time_to_next_midnight_us = g_date_time_difference (next_midnight, now);
+
+  priv->dispatch_aggregate_timers_daily_source_id =
+    eins_boottimeout_add_useconds (time_to_next_midnight_us,
+                                   clock_ticked_midnight_cb,
+                                   self);
+
+  priv->current_aggregate_tally_date = g_date_time_ref (now);
+}
+
+static gboolean
+on_timer_stopped_cb (EmerAggregateTimer     *timer,
+                     GDBusMethodInvocation  *invocation,
+                     EmerAggregateTimerImpl *timer_impl)
+{
+  EmerDaemon *self = g_object_get_data (G_OBJECT (timer_impl), "daemon");
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+  g_autoptr(GDateTime) now = NULL;
+  g_autoptr(GError) error = NULL;
+  gint64 now_monotonic_us;
+
+  now = g_date_time_new_now_local ();
+  now_monotonic_us = g_get_monotonic_time ();
+  emer_aggregate_timer_impl_stop (timer_impl, now, now_monotonic_us, &error);
+  if (error)
+    g_warning ("Error stopping timer: %s", error->message);
```

Comment:
You could return this error to the caller

Reply:
Good idea